server/src/main/java/org/elasticsearch/search/rank/RankSearchContext.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.rank;
+
+import org.apache.lucene.search.Query;
+import org.elasticsearch.search.aggregations.SearchContextAggregations;
+import org.elasticsearch.search.internal.FilteredSearchContext;
+import org.elasticsearch.search.internal.SearchContext;
+import org.elasticsearch.search.profile.Profilers;
+import org.elasticsearch.search.query.QueryPhase;
+import org.elasticsearch.search.query.QuerySearchResult;
+import org.elasticsearch.search.suggest.SuggestionSearchContext;
+
+import java.io.IOException;
+
+/**
+ * Manages the appropriate values when executing multiple queries
+ * on behalf of ranking.
+ *
+ * If the rank query is not set, this will behave like a default search
+ * context for aggregations and hit tracking with the important exception
+ * that size is set to [0]. This allows that query to be run without scoring.
+ *
+ * The rank query is set for each query in the {@link QueryPhase} and executed
+ * a single time with new results. The results must be collected out of
+ * {@link QuerySearchResult} after each execution.
+ */
+public class RankSearchContext extends FilteredSearchContext {
```

Comment:
I was wondering if we could use the original search context directly for the initial aggs search and this one only for the different rank searches. I feel like that would remove the null checks below and simplify things a bit.

For instance,  the aggregations method could simply throw, as well as all the features that the rank searches don't support (yet).

Reply:
I went ahead and just outright removed the unused methods altogether. I think that's simpler than trying to throw for the unused methods.